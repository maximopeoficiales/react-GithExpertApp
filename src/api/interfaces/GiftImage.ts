export default interface GiftImage {
    id: string,
    title: string,
    bitly_url: string,
    imgUrl: string,
}