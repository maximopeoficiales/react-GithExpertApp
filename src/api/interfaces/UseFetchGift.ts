import GiftImage from "./GiftImage";

export default interface UseFetchGift {
    gifts: GiftImage[],
    loading: boolean,
}