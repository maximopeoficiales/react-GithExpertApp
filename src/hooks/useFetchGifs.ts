import { useEffect, useState } from "react";
import UseFetchGift from "../api/interfaces/UseFetchGift";
import getGiftByCategory from "../helpers/getGift";

const useFetchGifs = (category: string): UseFetchGift => {

    const [gifts, setGifts] = useState({
        gifts: [], loading: true
    } as UseFetchGift);

    useEffect(() => {
        (async () => {
            setGifts({
                gifts: await getGiftByCategory(category),
                loading: false
            } as UseFetchGift);
        })();
    }, [category])
    return gifts;
}
export default useFetchGifs;