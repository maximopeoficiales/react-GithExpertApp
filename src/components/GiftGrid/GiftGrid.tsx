import React from "react";
import GiftImage from "../../api/interfaces/GiftImage";
import GiftGridItem from "../GiftGridItem/GiftGridItem";
import useFetchGifs from "../../hooks/useFetchGifs";
interface MyProps {
  category: string;
}
const defaultProps = {};

const GiftGrid = (props: MyProps) => {
  props = { ...defaultProps, ...props };
  const { category } = props;
  const { gifts, loading } = useFetchGifs(category);

  return (
    <div data-testid="GiftGrid" className="container-grid">
      <h1 className="title-category">{category}</h1>
      <p className="text-loading">{loading ? "Cargando...." : ""}</p>
      <div className="card-grid">
        {gifts.map((e: GiftImage) => (
          <GiftGridItem key={e.id} image={e} />
        ))}
      </div>
    </div>
  );
};

GiftGrid.propTypes = {};

export default GiftGrid;
