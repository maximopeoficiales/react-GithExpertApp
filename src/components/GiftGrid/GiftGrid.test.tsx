import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import GiftGrid from "./GiftGrid";
import { shallow } from "enzyme";
import useFetchGifs from "../../hooks/useFetchGifs";
import UseFetchGift from "../../api/interfaces/UseFetchGift";
import GiftImage from "../../api/interfaces/GiftImage";

jest.mock("../../hooks/useFetchGifs");

describe("<GiftGrid />", () => {
  const category = "Pokemon";

  const gifts: GiftImage[] = [
    {
      id: "43",
      imgUrl: "alguna url",
      title: "algun titulo",
      bitly_url: "algun link",
    },
  ];
  test("debe de mostrar el snapshot <GiftGrid/>", () => {
    // su estado inicial de useFetchGifs es en true y sin data
    (useFetchGifs as jest.Mock).mockReturnValue({
      gifts: [],
      loading: true,
    } as UseFetchGift);
    const wrapper = shallow(<GiftGrid category={category} />);
    // verifica si se renderizo correctamente
    expect(wrapper).toMatchSnapshot();
  });

  test("debe mostrar items cuando se cargan las images useFetchGifs <GiftGrid/>", () => {
    // su estado inicial de useFetchGifs es en false y con data
    (useFetchGifs as jest.Mock).mockReturnValue({
      gifts,
      loading: false,
    } as UseFetchGift);
    const wrapper = shallow(<GiftGrid category={category} />);
    // verifica si se renderizo correctamente
    expect(wrapper).toMatchSnapshot();
    // si el p  esta vacio
    expect(wrapper.find("p").text().trim()).toBe("");
    expect(wrapper.find("GiftGridItem").length).toBe(gifts.length);
  });
});
