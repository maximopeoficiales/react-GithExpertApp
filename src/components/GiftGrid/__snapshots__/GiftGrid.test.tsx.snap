// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GiftGrid /> debe de mostrar el snapshot <GiftGrid/> 1`] = `
<div
  className="container-grid"
  data-testid="GiftGrid"
>
  <h1
    className="title-category"
  >
    Pokemon
  </h1>
  <p
    className="text-loading"
  >
    Cargando....
  </p>
  <div
    className="card-grid animate__animated animate__backInLeft"
  />
</div>
`;

exports[`<GiftGrid /> debe mostrar items cuando se cargan las images useFetchGifs <GiftGrid/> 1`] = `
<div
  className="container-grid"
  data-testid="GiftGrid"
>
  <h1
    className="title-category"
  >
    Pokemon
  </h1>
  <p
    className="text-loading"
  />
  <div
    className="card-grid animate__animated animate__backInLeft"
  >
    <GiftGridItem
      image={
        Object {
          "bitly_url": "algun link",
          "id": "43",
          "imgUrl": "alguna url",
          "title": "algun titulo",
        }
      }
      key="43"
    />
  </div>
</div>
`;
